crates/nu-engine/src/closure_eval.rs

```diff
@@ -0,0 +1,236 @@
+/// so nothing needs to be done following [`ClosureEval::run_with_input`] or [`ClosureEval::run_with_value`].
+pub struct ClosureEval {
+    engine_state: EngineState,
+    stack: Stack,
+    block: Arc<Block>,
+    arg_index: usize,
+    env_vars: Vec<EnvVars>,
+    env_hidden: HashMap<String, HashSet<String>>,
+    eval: EvalBlockWithEarlyReturnFn,
+}
+
+impl ClosureEval {
+    /// Create a new [`ClosureEval`].
+    pub fn new(engine_state: &EngineState, stack: &Stack, closure: Closure) -> Self {
+        let engine_state = engine_state.clone();
+        let stack = stack.captures_to_stack(closure.captures);
+        let block = engine_state.get_block(closure.block_id).clone();
+        let env_vars = stack.env_vars.clone();
+        let env_hidden = stack.env_hidden.clone();
+        let eval = get_eval_block_with_early_return(&engine_state);
+
+        Self {
+            engine_state,
+            stack,
+            block,
+            arg_index: 0,
+            env_vars,
+            env_hidden,
+            eval,
+        }
+    }
+
+    /// Sets whether to enable debugging when evaluating the closure.
+    ///
+    /// By default, this is controlled by the [`EngineState`] used to create this [`ClosureEval`].
+    pub fn debug(&mut self, debug: bool) -> &mut Self {
+        self.eval = eval_fn(debug);
+        self
+    }
+
+    fn try_add_arg(&mut self, value: Cow<Value>) {
+        if let Some(var_id) = self
+            .block
+            .signature
+            .get_positional(self.arg_index)
+            .and_then(|var| var.var_id)
+        {
+            self.stack.add_var(var_id, value.into_owned());
+            self.arg_index += 1;
+        }
+    }
+
+    /// Add an argument [`Value`] to the closure.
+    ///
+    /// Multiple [`add_arg`](Self::add_arg) calls can be chained together,
+    /// but make sure that arguments are added based on their positional order.
+    pub fn add_arg(&mut self, value: Value) -> &mut Self {
+        self.try_add_arg(Cow::Owned(value));
+        self
+    }
+
+    /// Run the closure, passing the given [`PipelineData`] as input.
+    ///
+    /// Any arguments should be added beforehand via [`add_arg`](Self::add_arg).
+    pub fn run_with_input(&mut self, input: PipelineData) -> Result<PipelineData, ShellError> {
+        self.arg_index = 0;
+        self.stack.with_env(&self.env_vars, &self.env_hidden);
+        (self.eval)(&self.engine_state, &mut self.stack, &self.block, input)
+    }
+
+    /// Run the closure using the given [`Value`] as both the pipeline input and the first argument.
+    ///
+    /// Using this function after or in combination with [`add_arg`](Self::add_arg) is most likely an error.
+    /// This function is equivalent to `self.add_arg(value)` followed by `self.run_with_input(value.into_pipeline_data())`.
+    pub fn run_with_value(&mut self, value: Value) -> Result<PipelineData, ShellError> {
+        self.try_add_arg(Cow::Borrowed(&value));
```

Comment:
Is there a reason for the Borrowed? It seems that it will get converted `.to_owned()` inside `try_add_arg()` anyway.

I'm asking because I'd prefer having to call `.add_arg()` and `.run_with_input()` explicitly to remove the possibility of using this method wrong, as you noted in the comment above. But if the Borrowed indeed plays a role, we can keep it, it just seems that it's not necessary.

Reply:
If a command populates the `$in` value, then the closure's first parameter is most likely optional. Some contrived examples:
```nushell
ls | each { values }
ls | group-by { get type }
```

Previously, every command that did this had the following code:
```rust
if let Some(var) = block.signature.get_positional(0) {
    if let Some(var_id) = &var.var_id {
        stack.add_var(*var_id, value.clone());
    }
}
```

And this would only clone `value` if necessary. So, to avoid performance regressions, I added `run_with_value` and the whole `Cow` workaround, since I think a lot of commands have this form. Then again, I didn't do any benchmarks, so this may not actually matter.

I could add an intermediate builder type that is returned from `add_arg` and remove the `run_with_value` function for this type, but this might add some more code.

Reply:
OK, then if there would be a potential regression, we can keep the Borrowed.